### DISCUSSION TITLE
Offtopic: need hints for GitHub 2FA [SOLVED]

### DISCUSSION THREAD

**Discussion by markus-oberhumer** [upx/upx#704] [Q&A] [answered]
GitHub has informed me that I will need to activate 2FA. Well,
- I refuse to use a proprietary OS (aka smartphone)
- I refuse to use proprietary software/services

Any hints how to accomplish 2FA in a truly Open Source manner?

### EDIT 1: SOLVED, THANKS TO EVERYONE!

Use `2fa` from https://github.com/rsc/2fa . See below.

### EDIT 2:  EVEN SIMPLER

I can only mark a single answer, but @bitraid provided the info that `oathtool --totp --base32 <your_two_factor_secret>` is basically all you need.

**Comment by bitraid** [comment]
You could use [pass-otp](https://github.com/tadfisher/pass-otp), which is a [pass](https://www.passwordstore.org/) extension, or you could use [pass and oathtool](https://blog.separateconcerns.com/2020-04-03-2fa-pass-oathtool.html).

**Comment by jreiser** [comment]
What about FIDO2 USB hardware for 2FA,  such as various Yubikey models, and other manufacturers, too?

**Comment by markus-oberhumer** [comment]
@bitraid thanks, I will try these on the weekend!

@jreiser good idea, this might be an option if the other solutions prove too inconvenient!

**Comment by jreiser** [comment]
Portability: I use 6 different machines (2 Fedora, 2 Debian, 2 Macs (El Capitan: OS X 10.11.6 x86_64, Ventura: MacOS 13.4.1 Apple M1) so it would be nice to support them all.  (I suppose `homebrew` might work for the Macs.)

**Comment by mgrinzPlayer** [comment]
TOTP, I first tried it with KeePassXC. There are many offline calculators and I tried few of them from here:
- [Pascal](https://github.com/search?q=totp+language%3Apascal&type=repositories)
- [Bash](https://github.com/search?q=totp+language%3Abash&type=repositories)
- [javascript](https://github.com/search?q=totp+language%3Ajavascript&type=repositories)

**Reply by mgrinzPlayer** [reply]
>I'll keep this discussion open for now

Just an example in Pascal ("Free Pascal/Lazarus", standard installation, vanilla):
```Pascal
//add DateUtils and HMAC
uses
  Classes, SysUtils, Forms, StdCtrls, DateUtils, HMAC;
var
  validBase32Chars:string='ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';

function base32decode(input: string):string;
var i,j,n,p: integer;
    output: string='';
begin
  input:=input.ToUpper;
  j:=0;
  n:=0;
  for i:=1 to Length(input) do
  begin
    p:=Pos(input[i],validBase32Chars);
    if p>0 then begin
      j:=j+5;
      n:=(n shl 5)+(p-1);
    end;
    if j>=8 then begin
      j:=j-8;
      output:=output+char(n shr j);
      n:=n and ((1 shl j)-1);
    end;
  end;
  exit(output);
end; 

procedure showTOTP()
var
    i: Integer;
    TOTP_int64: Int64;
    TOTP_string: string;
    seconds_left: Integer;

    secret_decoded: string='';
    timeSegmentStr: string='';

    timeSegment: Int64;
    buf: pchar;
    digest: THMACSHA1Digest;
begin
  secret_decoded:=base32decode('GitHub 2FA setup key here');

  // alternative
  //secret_decoded:=base32decode(secret_base32.Text);

  timeSegment:=SwapEndian(DateTimeToUnix(Now,false) div 30);
  buf:=@timeSegment;
  for i:=0 to 7 do timeSegmentStr:=timeSegmentStr+buf[i];

  digest:=HMACSHA1Digest(secret_decoded,timeSegmentStr);

  i:=digest[19] and $f;
  TOTP_int64:=Int64((digest[i+0] and $7F) shl 24) or
              (digest[i+1] shl 16) or
              (digest[i+2] shl 8) or
              (digest[i+3]);

  TOTP_string:=Format('%.6d',[TOTP_int64 mod 1000000]);
  seconds_left:=30-(DateTimeToUnix(Now,false) mod 30);

  Memo1.Append(Format('TOTP: %s',  [TOTP_string]  ));
  Memo1.Append(Format('Seconds left: %d',  [seconds_left]  ));
  //or use writeln or something...
end;
```
Just be sure your machine has correct date and time set.

**Comment by 239** [comment]
I'm also using FIDO2 security keys, not only for 2FA - most secure option.
For Android I can recommend [Aegis](https://getaegis.app), which can also be used on Google-free devices.
On all Linux desktops I'm using [Authenticator](https://flathub.org/apps/com.belmoussaoui.Authenticator), it can import Aegis exports.

**Comment by markus-oberhumer** [comment]
@239 No smartphone, so no Aegis. Looks like Authenticator is a Gnome app, I'll try installing that in a container.

**Reply by 239** [reply]
Since TOTP needs no connectivity it can be used offline on any device.
Not sure if Flatpak apps will run in containers, CLI alternatives are probably better.

**Reply by jreiser** [reply]
`Since TOTP needs no connectivity it can be used offline on any device.`
This cannot be true, because "Time-limited" requires a clock or a timer.  A clock is an external device , or [eventually] requires an external power source; and a timer needs an external connection in order to set it (or get the information to set it.)

**Reply by 239** [reply]
I meant that no network connection is required. You can set the clock manually, just like in the good old days. One of my 2FA-backups is an old smartphone which works offline.

**Comment by markus-oberhumer** [comment]
### Solved, thanks for all your help! A quick summary:

I quickly tried `pass-otp`, but it wanted me to mess around with GPG keys and the docs didn't help either.

Fortunately I then discovered `keyfob` https://github.com/StevenACoffman/keyfob (which is based on `2fa` https://github.com/rsc/2fa/ )

Things are completely trivial:
```
keyfob add github your-two-factor-secret-from-the-github-popup-window
keyfob otp github
```

### A closer look

Wondering where `keyfob` actually stores its data I discovered the wonderful world of Gnome Keyrings. Here `lssecret` https://gitlab.com/GrantMoyer/lssecret helped me to understand I can use the standard

```
secret-tool search --all xdg:schema org.freedesktop.Secret.Generic
```

On my machine the actual data seem to be stored in `~/.local/share/keyrings`.

(This seems an overkill, I'll probably try `2fa` )

**Reply by jreiser** [reply]
`your-two-factor-secret-from-the-github-popup-window`
What is the URL of the page which presents this popup?  The GitHub docs assume that you already understand the processes and the pitfalls, and already know the answer to your question.  I probed the Settings for my profile, but did not see any popup.

**Reply by markus-oberhumer** [reply]
See "Unable to scan?" and click on `setup key`.

BTW, I've now switched to `2fa` - see build instructions below.

![Screenshot_2023-08-31_20-13-37](https://github.com/upx/upx/assets/2130496/c3401b0b-e2f2-42da-805a-55f344177c26)

**Comment by markus-oberhumer** [comment, marked as answer]
### Update: 

`2fa` is much simpler, and it stores the keys in a file `~/.2fa` which can get easily copied around.

```
$ git clone https://github.com/rsc/2fa
$ cd 2fa
$ go mod vendor
$ go build .
```

(I can't believe how much time I spent on what seems a quite trivial problem.)

**Comment by bitraid** [comment]
@markus-oberhumer `pass` uses a GPG key to encrypt its password store contents (`~/.password-store`). You can can create a new key or use an existing one: `pass init <public_key>` (you can run `gpg --list-keys` to view available keys). Then run `pass otp insert github` and paste `otpauth://totp/GitHub/<user_name>?secret=<setup_key>&issuer=GitHub` (replace `<user_name>` and `<setup_key>`). And that's all - after that you just run `pass otp github` to get the code.

**Comment by markus-oberhumer** [comment]
@bitraid Yes, but all this GPG/Keyring  stuff is an annoying pseudo-security. I only need to trust the filesystem & permission, just like with my ssh keys.

**Comment by bitraid** [comment]
I password protect my gpg/ssh keys for actual security.

**Reply by markus-oberhumer** [reply]
Sure, but to prevent brute force attacks you will need **very** good (and possibly lots) different passwords that you cannot remember.

So where do you store those meta-passwords? :smile:

**Reply by bitraid** [reply]
I only have to remember one master password that unlocks the password store - then all the other passwords are there!

**Reply by bitraid** [reply]
If you don't care about encryption, you could just run (or make an alias to) `oathtool --totp --base32 <setup_key>`.

**Reply by markus-oberhumer** [reply]
> If you don't care about encryption, you could just run (or make an alias to) `oathtool --totp --base32 <setup_key>`.

Thanks, I wish I had known that from the start!

**Reply by bitraid** [reply]
Yeah, sorry. I should have explained in my first reply about pass and oathtool, that pass is only needed for encryption.

**Comment by markus-oberhumer** [comment]
I'll keep this discussion open for now, maybe someone else might find this useful.

**Comment by jreiser** [comment]
**Missing**: enumerate **all** the failure modes (or soon-to-become-a-failure modes), and the recovery procedure (if any) for each failure mode.  This includes all the timeouts, all the typos or other incorrect responses, all the `lost connection`, all the bit-level corruptions of the communication channels, all the crashes and bugs of any software (local or remote), etc. The consequences of failure may be severe (including permanent, non-recoverable loss of the GitHub user ID), and in many cases using GItHub 2FA is a first-time experience with the tools and/or the process.

**Reply by markus-oberhumer** [reply]
Yeah it's annoying, but then what would you do if you were in GitHub's position? The problem is that security is hard - it's hard even for people who have some understanding of the actual details, so it's consequential that most normal users don't care too much - which then causes problems in the first place...

**Reply by jreiser** [reply]
`What would you do?`  As they occur, create a catalog of all known failures: every time a GitHub user ID is canceled or abandoned because of security, etc.  Investigate each one: query the user about the circumstances, etc.  Categorize the reasons. Ask the experts to diagnose, explain, give the remedies and recommend best practices.  Publish.

**Reply by bitraid** [reply]
GitHub provides 2FA for many years now (and so do many other services). When enabled, the user is responsible to have their recovery codes stored somewhere safe. And that is for event that the user loses access to the device that does the authentication - I don't think there are cases where users are locked out of their accounts because of network/software problems, otherwise we would hear about it (people would furiously complain about it online). I would also guess that most people would prefer to use the mobile app instead of an authenticator app, which is more simple, as there is no need for TOTP (the authentication is performed by receiving a push notification, aka push authentication).

**Comment by mgrinzPlayer** [comment]
If anyone interested, as of this writing, GitHub accepts Time-based codes from -1.5 minutes up to +1.5 minutes range.

e.g. assume you send code at 12:30:02 (we are at 12:30:00 - 12:30:29.99... interval)

Codes generated at interval:
- 12:28:00 - 12:28:29.99... **NOT** accepted, too old
- 12:28:30 - 12:28:59.99... be accepted
- 12:29:00 - 12:29:29.99... be accepted
- 12:29:30 - 12:29:59.99... be accepted
- 12:30:00 - 12:30:29.99... be accepted (current interval)
- 12:30:30 - 12:30:59.99... be accepted
- 12:31:00 - 12:31:29.99... be accepted
- 12:31:30 - 12:31:59.99... be accepted
- 12:31:30 - 12:31:59.99... **NOT** accepted, too new (just wait 30 seconds and it will work)

Tested it right now, with `((DateTimeToUnix(Now,false) div 30) +X )` [used in this code](https://github.com/upx/upx/discussions/704#discussioncomment-6888916) (where X was -5, -4, -3, 1, 2, 3, 4)

Your machine internal clock doesn't have to be ±0.001 seconds.
±60 seconds would be enough. 

**PS**:
if code is too old or too new, GitHub will respond with `Two-factor authentication failed` message.

if code was already used, it will respond with `The two-factor code you entered has already been used or is too old to be used.`
(that "too old" part is slightly misleading, because too old codes just give "Two-factor authentication failed" even those already used)

if you play too fast with testing 2FA, GitHub will throw
```
"We were unable to authenticate your request because too many codes have been submitted.
Please wait a few minutes and contact support if you continue to have problems."
```

This is why I tested "-5, -4, ..." thing with longer breaks.

**Reply by markus-oberhumer** [reply]
From what I understand TOTP is based on 30 second intervals. From a provider side it makes sense to allow +x/-x seconds deviation to not  unnecessarily annoy/confuse your users.

**Reply by mgrinzPlayer** [reply]
Yes.

In recent days I read first few pages from this search [link](https://github.com/search?q=2FA&type=discussions) (+some google searches) and I saw assumptions it is "one and half times interval" - ±45 seconds, I also saw -60/+15 version. I also see a lot of panic: "help me migrate my source code", "another app on my phone", "they want my phone number". Yes, GH 'Your GitHub account, will soon require 2FA' email suggest to use your smartphone.

I was curious how GH deals with it. I decided to not touch the clock, just simulate time deviation, and try "`((DateTimeToUnix(Now,false) div 30) +X )` generated code" at the beginning of 30 seconds cycle (I pasted generated code about 2 seconds after beginning of cycle).

The exact -X/+Y, for me it is -90/+90. Of course they can change it later.
